wicket/src/ui/panes/update.rs

```diff
@@ -2393,51 +2395,87 @@ fn all_installed_versions(
     }
 }
 
+type ArtifactVersions = Vec<(SemverVersion, Option<Vec<u8>>)>;
+
 fn artifact_version(
     id: &ComponentId,
-    component: UpdateComponent,
-    versions: &BTreeMap<KnownArtifactKind, SemverVersion>,
+    update_component: UpdateComponent,
+    versions: &BTreeMap<KnownArtifactKind, ArtifactVersions>,
+    inventory: &Inventory,
 ) -> String {
-    let artifact = match (id, component) {
+    let (artifact, multiple) = match (id, update_component) {
         (ComponentId::Sled(_), UpdateComponent::RotBootloader) => {
-            KnownArtifactKind::GimletRotBootloader
+            (KnownArtifactKind::GimletRotBootloader, true)
         }
         (ComponentId::Sled(_), UpdateComponent::Rot) => {
-            KnownArtifactKind::GimletRot
+            (KnownArtifactKind::GimletRot, true)
         }
         (ComponentId::Sled(_), UpdateComponent::Sp) => {
-            KnownArtifactKind::GimletSp
+            (KnownArtifactKind::GimletSp, false)
         }
         (ComponentId::Sled(_), UpdateComponent::Host) => {
-            KnownArtifactKind::Host
+            (KnownArtifactKind::Host, false)
         }
         (ComponentId::Switch(_), UpdateComponent::RotBootloader) => {
-            KnownArtifactKind::SwitchRotBootloader
+            (KnownArtifactKind::SwitchRotBootloader, true)
         }
         (ComponentId::Switch(_), UpdateComponent::Rot) => {
-            KnownArtifactKind::SwitchRot
+            (KnownArtifactKind::SwitchRot, true)
         }
         (ComponentId::Switch(_), UpdateComponent::Sp) => {
-            KnownArtifactKind::SwitchSp
+            (KnownArtifactKind::SwitchSp, false)
         }
         (ComponentId::Psc(_), UpdateComponent::RotBootloader) => {
-            KnownArtifactKind::PscRotBootloader
+            (KnownArtifactKind::PscRotBootloader, true)
         }
         (ComponentId::Psc(_), UpdateComponent::Rot) => {
-            KnownArtifactKind::PscRot
+            (KnownArtifactKind::PscRot, true)
+        }
+        (ComponentId::Psc(_), UpdateComponent::Sp) => {
+            (KnownArtifactKind::PscSp, false)
         }
-        (ComponentId::Psc(_), UpdateComponent::Sp) => KnownArtifactKind::PscSp,
 
         // Switches and PSCs do not have a host.
         (ComponentId::Switch(_), UpdateComponent::Host)
         | (ComponentId::Psc(_), UpdateComponent::Host) => {
             return "N/A".to_string()
         }
     };
-    versions
-        .get(&artifact)
-        .cloned()
-        .map_or_else(|| "UNKNOWN".to_string(), |v| v.to_string())
+    match versions.get(&artifact) {
+        None => "UNKNOWN".to_string(),
+        Some(v) => {
+            let component = match inventory.get_inventory(id) {
+                Some(c) => c,
+                None => return "UNKNOWN".to_string(),
+            };
+            let cnt = v.len();
+            for (vers, sign) in v {
+                match (sign, component.rot_sign()) {
```

Comment:
I don't understand the `for`/`match` here; it looks like it could return different values depending on the ordering of `v`, maybe? E.g., if v has two entries, one with a sign and one without, and `component.rot_sign()` is none, will we either return `vers` (line 2455) if the "without a sign" is first or `UNKNOWN (MISSING SIGN)` (line 2461) if the "with a sign" is first?

Reply:
The assumption is that either all entries will have a `SIGN` or none of them. This should be checked elsewhere or handled/documented better. This is also related to the type you pointed out in the `Vec` which is very much a theme here and something I need to rework.

Reply:
I added a few more comments. It's not possible for us to construct a repository with a mixture of `SIGN` and no `SIGN` because we now require it when reading RoT hubris archives and we only populate the sign field from RoT images.

Reply:
I added test to verify that we fail if we try and read a caboose with no `SIGN` on the RoT